File: server/src/lib/license.js
'use strict';

/**
 *  صدور مجوز اشتراک — همان توکنی که برنامه امضایش را می‌سنجد.
 *
 *  چرا اصلاً مجوز؟ برنامه باید آفلاین هم بداند اشتراک باز است یا نه.
 *  اگر جواب «باز است» فقط یک پرچم در حافظه‌ی گوشی باشد، هر کسی می‌تواند
 *  عوضش کند. پس سرور یک برگه‌ی امضاشده می‌دهد و برنامه فقط امضا را
 *  می‌سنجد: بدون کلید خصوصی سرور، ساختن چنین برگه‌ای شدنی نیست.
 *
 *  توکن سه بخش دارد — `header.payload.signature` — با ES256 روی منحنی
 *  P-256، کدگذاری base64url. دقیقاً همان چیزی که `License.kt` می‌خواند.
 *
 *  دو قید که مجوز دزدیده‌شده را بی‌ارزش می‌کنند:
 *    • به **دستگاه** بسته است (`dev`)؛ کپی کردنش روی گوشی دیگر کار نمی‌کند.
 *    • عمر کوتاهی دارد (`exp`)، کوتاه‌تر از خود اشتراک. پس برنامه هر چند
 *      روز باید دوباره از سرور بگیرد و لغو اشتراک زود اثر می‌کند. تا آن
 *      وقت هم آفلاین کار می‌کند.
 *
 *  کلید یک بار ساخته می‌شود و در `app_config` می‌ماند. اگر
 *  `LICENSE_PRIVATE_KEY` در محیط باشد همان استفاده می‌شود — برای وقتی که
 *  چند سرور باید یک کلید داشته باشند.
 */

const crypto = require('crypto');
const plans = require('./plans');

const ISSUER = 'tohid-license-server';
const AUDIENCE = 'tohid-shop-app';

/** عمر خود مجوز (نه اشتراک): ده روز */
const TOKEN_TTL_MS = 10 * 24 * 60 * 60 * 1000;

const KEY_PRIVATE = 'license_private_key';
const KEY_PUBLIC = 'license_public_key';

let cached = null;

/**
 *  جفت‌کلید سرور. یک بار ساخته و ذخیره می‌شود.
 *
 *  کلید عمومی هم ذخیره می‌شود تا برای هر درخواست از روی خصوصی حساب نشود.
 */
async function keys() {
  if (cached) return cached;

  const fromEnv = process.env.LICENSE_PRIVATE_KEY;
  if (fromEnv && fromEnv.trim()) {
    const privateKey = crypto.createPrivateKey(fromEnv.trim());
    cached = { privateKey, publicSpki: spkiOf(privateKey) };
    return cached;
  }

  const stored = await plans.getConfig(KEY_PRIVATE, '');
  if (stored) {
    const privateKey = crypto.createPrivateKey(stored);
    const publicSpki = (await plans.getConfig(KEY_PUBLIC, '')) || spkiOf(privateKey);
    cached = { privateKey, publicSpki };
    return cached;
  }

  const pair = crypto.generateKeyPairSync('ec', {
    namedCurve: 'prime256v1',
    privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
    publicKeyEncoding: { type: 'spki', format: 'pem' },
  });
  const privateKey = crypto.createPrivateKey(pair.privateKey);
  const publicSpki = spkiOf(privateKey);

  await plans.setConfig(KEY_PRIVATE, pair.privateKey);
  await plans.setConfig(KEY_PUBLIC, publicSpki);

  cached = { privateKey, publicSpki };
  return cached;
}

/** کلید عمومی به شکل base64 از SPKI DER — همان چیزی که برنامه می‌خواهد */
function spkiOf(privateKey) {
  return crypto
    .createPublicKey(privateKey)
    .export({ type: 'spki', format: 'der' })
    .toString('base64');
}

async function publicKey() {
  return (await keys()).publicSpki;
}

function b64url(buf) {
  return Buffer.from(buf).toString('base64')
    .replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/**
 *  یک مجوز امضاشده.
 *
 *  @param {object} p
 *  @param {string} p.deviceUid   شناسه‌ی دستگاه — مجوز فقط روی همین کار می‌کند
 *  @param {string} p.accountId   شناسه‌ی حساب
 *  @param {string[]} p.features  قابلیت‌های باز
 *  @param {string[]} p.core      قابلیت‌هایی که همیشه بازند
 *  @param {number} p.subscriptionEndsAt پایان اشتراک (میلی‌ثانیه)
 */
async function issue({
  deviceUid,
  deviceName = '',
  accountId,
  features = [],
  core = [],
  subscriptionEndsAt = 0,
  plan = '',
  planTitle = '',
  at = Date.now(),
}) {
  const { privateKey } = await keys();

  const header = { alg: 'ES256', typ: 'TLIC' };

  // مجوز از خود اشتراک زودتر تمام می‌شود تا لغو اشتراک زود اثر کند —
  // ولی نه دیرتر از پایان اشتراک، وگرنه چند روز مجانی می‌دهد
  const expiresAt = Math.min(at + TOKEN_TTL_MS, subscriptionEndsAt || at + TOKEN_TTL_MS);

  /*
   * نام فیلدها همان چیزی است که برنامه می‌خواند — نه چیز دیگری.
   *
   * تا امروز اینجا `dev` و `acc` نوشته می‌شد، ولی هر دو کلاینت
   * (اندروید و وب) `duid` و `sub` را می‌خوانند. یعنی `payload.duid`
   * همیشه undefined بود و مجوز با `device_mismatch` رد می‌شد:
   * **اشتراک پولی روی هیچ دستگاهی فعال نمی‌شد.**
   *
   * تست‌های کلاینت این را نمی‌گرفتند چون خودشان مجوز می‌ساختند و
   * همان نام‌های درست را می‌گذاشتند؛ هیچ تستی مجوزِ واقعیِ سرور را با
   * چشمِ کلاینت نگاه نمی‌کرد. حالا می‌کند (`license-format.test.js`).
   *
   * `dev` و `acc` هم کنارشان می‌مانند تا اگر ابزاری بیرون از این مخزن
   * آن‌ها را می‌خواند، نشکند.
   */
  const payload = {
    iss: ISSUER,
    aud: AUDIENCE,
    duid: deviceUid,
    sub: accountId,
    dev: deviceUid,
    dev_name: deviceName,
    acc: accountId,
    iat: at,
    nbf: at - 60_000,            // یک دقیقه ارفاق برای ساعت گوشی
    exp: expiresAt,
    sub_ends: subscriptionEndsAt,
    feat: features,
    core,
    plan,
    plan_title: planTitle,
  };

  const signing =
    `${b64url(JSON.stringify(header))}.${b64url(JSON.stringify(payload))}`;

  // امضای خام (r||s) نه DER: برنامه همین را می‌خواند
  const signature = crypto.sign(null, Buffer.from(signing, 'utf8'), {
    key: privateKey,
    dsaEncoding: 'ieee-p1363',
  });

  return {
    token: `${signing}.${b64url(signature)}`,
    expiresAt,
    subscriptionEndsAt,
  };
}

module.exports = { issue, publicKey, ISSUER, AUDIENCE, TOKEN_TTL_MS };

File: android-native/app/src/main/java/ir/vil3ntec/tohid/sync/License.kt
package ir.vil3ntec.tohid.sync

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.security.KeyFactory
import java.security.Signature
import java.security.spec.X509EncodedKeySpec

/**
 *  بررسیِ مجوزِ اشتراک.
 *
 *  مجوز یک توکنِ امضاشده است: `header.payload.signature` با ES256 روی
 *  منحنیِ P-256. بدونِ امضای معتبر هیچ چیزی پذیرفته نمی‌شود — نه تاریخ،
 *  نه قابلیت، نه نامِ پلن. وگرنه هرکسی می‌توانست با دست‌کاریِ یک فایلِ
 *  متنی، اشتراکِ همیشگی برای خودش بسازد.
 *
 *  دو نکتهٔ فنی که اشتباه‌شان کلِ بررسی را بی‌اثر می‌کند:
 *    • امضا خام است (r و s پشتِ سرِ هم، ۶۴ بایت) ولی جاوا فقط DER
 *      می‌فهمد، پس باید تبدیل شود. اگر تبدیل نشود، جاوا امضا را خراب
 *      می‌داند و هیچ مجوزی هرگز قبول نمی‌شود.
 *    • کدگذاری base64url است، نه base64 معمولی.
 */
object License {

  private const val ISSUER = "tohid-license-server"
  private const val AUDIENCE = "tohid-shop-app"

  data class Payload(
    val deviceUid: String,
    val accountId: String,
    val issuedAt: Long,
    val notBefore: Long,
    val expiresAt: Long,
    val subscriptionEndsAt: Long,
    val features: List<String>,
    val core: List<String>,
    val plan: String?,
    val planTitle: String?,
  )

  sealed interface Verdict {
    data class Valid(val payload: Payload) : Verdict
    /** دلیل به همان کدهای نسخهٔ وب: format | header | signature | issuer | audience | device_mismatch */
    data class Invalid(val reason: String) : Verdict
  }

  /** وضعیتِ اشتراک در همین لحظه */
  enum class State { NONE, INVALID, PENDING, ACTIVE, GRACE, EXPIRED }

  data class Status(
    val state: State,
    val payload: Payload? = null,
    val reason: String? = null,
  ) {
    val features: List<String> get() = if (state == State.ACTIVE || state == State.GRACE) payload?.features.orEmpty() else emptyList()
  }

  /**
   * امضا و ساختارِ مجوز.
   *
   * @param publicKeySpki کلیدِ عمومیِ سرور، base64 از SPKI DER
   * @param deviceUid شناسهٔ همین دستگاه — مجوزِ دستگاهِ دیگری پذیرفته نمی‌شود
   */
  fun verify(token: String, publicKeySpki: String, deviceUid: String): Verdict {
    val parts = token.split('.')
    if (parts.size != 3) return Verdict.Invalid("format")

    val header: JsonObject
    val payload: JsonObject
    try {
      header = parser.parseToJsonElement(String(decodeUrl(parts[0]), Charsets.UTF_8)).jsonObject
      payload = parser.parseToJsonElement(String(decodeUrl(parts[1]), Charsets.UTF_8)).jsonObject
    } catch (e: Exception) {
      return Verdict.Invalid("format")
    }

    if (header.text("alg") != "ES256" || header.text("typ") != "TLIC") {
      return Verdict.Invalid("header")
    }

    val signed = "${parts[0]}.${parts[1]}".toByteArray(Charsets.UTF_8)
    val ok = try {
      val key = KeyFactory.getInstance("EC")
        .generatePublic(X509EncodedKeySpec(decodeBase64(publicKeySpki.trim())))
      Signature.getInstance("SHA256withECDSA").run {
        initVerify(key)
        update(signed)
        verify(rawToDer(decodeUrl(parts[2])))
      }
    } catch (e: Exception) {
      return Verdict.Invalid("crypto_error")
    }
    if (!ok) return Verdict.Invalid("signature")

    // امضا درست است؛ حالا ببینیم مالِ همین برنامه و همین دستگاه است
    if (payload.text("iss") != ISSUER) return Verdict.Invalid("issuer")
    if (payload.text("aud") != AUDIENCE) return Verdict.Invalid("audience")
    if (payload.text("duid") != deviceUid) return Verdict.Invalid("device_mismatch")

    return Verdict.Valid(
      Payload(
        deviceUid = payload.text("duid"),
        accountId = payload.text("sub"),
        issuedAt = payload.number("iat"),
        notBefore = payload.number("nbf"),
        expiresAt = payload.number("exp"),
        subscriptionEndsAt = payload.number("sub_ends"),
        features = payload.list("feat"),
        core = payload.list("core"),
        plan = payload.text("plan").ifBlank { null },
        planTitle = payload.text("plan_title").ifBlank { null },
      )
    )
  }

  /**
   * وضعیت بر اساسِ مجوزِ ذخیره‌شده — همان حالت‌هایی که نسخهٔ وب داشت.
   * «مهلت» بازه‌ای است که اشتراک تمام شده ولی هنوز کار می‌کند.
   */
  fun status(token: String?, publicKeySpki: String?, deviceUid: String, now: Long): Status {
    if (token.isNullOrBlank() || publicKeySpki.isNullOrBlank()) return Status(State.NONE, reason = "no_license")

    return when (val v = verify(token, publicKeySpki, deviceUid)) {
      is Verdict.Invalid -> Status(State.INVALID, reason = v.reason)
      is Verdict.Valid -> {
        val p = v.payload
        when {
          now < p.notBefore -> Status(State.PENDING, p, "not_started")
          now > p.expiresAt -> Status(State.EXPIRED, p, "expired")
          now > p.subscriptionEndsAt -> Status(State.GRACE, p)
          else -> Status(State.ACTIVE, p)
        }
      }
    }
  }

  /* ------------------------------ ریزه‌کاری ------------------------------ */

  /*
   * از kotlinx خوانده می‌شود، نه org.json: آن یکی روی رایانه فقط یک
   * پوستهٔ خالی است و هر متدش خطا می‌دهد، پس بررسیِ مجوز — که مهم‌ترین
   * چیز برای سنجیدن است — اصلاً قابلِ آزمودن نمی‌شد.
   */
  private val parser = Json { ignoreUnknownKeys = true; isLenient = true }

  private fun JsonObject.text(key: String): String =
    (this[key] as? kotlinx.serialization.json.JsonPrimitive)
      ?.takeIf { it !is kotlinx.serialization.json.JsonNull }?.content.orEmpty()

  private fun JsonObject.number(key: String): Long =
    (this[key] as? kotlinx.serialization.json.JsonPrimitive)?.content?.toDoubleOrNull()?.toLong() ?: 0L

  private fun JsonObject.list(key: String): List<String> =
    (this[key] as? kotlinx.serialization.json.JsonArray)
      ?.mapNotNull { (it as? kotlinx.serialization.json.JsonPrimitive)?.content?.takeIf { s -> s.isNotBlank() } }
      .orEmpty()

  /** base64url — بدونِ padding، با `-` و `_` به‌جای `+` و `/` */
  fun decodeUrl(text: String): ByteArray = decodeBase64(text.replace('-', '+').replace('_', '/'))

  /**
   * base64، دستی.
   *
   * `java.util.Base64` از اندروید ۸ به بعد هست و این برنامه از اندروید ۷
   * کار می‌کند؛ `android.util.Base64` هم روی رایانه (هنگام آزمون) وجود
   * ندارد. بیست خط کد، هر دو مشکل را حل می‌کند.
   */
  fun decodeBase64(text: String): ByteArray {
    val clean = text.filter { it != '\n' && it != '\r' && it != ' ' && it != '=' }
    val out = java.io.ByteArrayOutputStream(clean.length * 3 / 4 + 3)
    var buffer = 0
    var bits = 0
    for (c in clean) {
      val v = ALPHABET.indexOf(c)
      require(v >= 0) { "نویسهٔ نامعتبر در base64" }
      buffer = (buffer shl 6) or v
      bits += 6
      if (bits >= 8) {
        bits -= 8
        out.write((buffer shr bits) and 0xFF)
      }
    }
    return out.toByteArray()
  }

  private const val ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"


  /**
   * امضای خام (r||s) به DER.
   *
   * هر عدد باید به‌صورت INTEGER با کمترین طول نوشته شود، و اگر بیتِ اولش
   * ۱ باشد یک بایتِ صفر جلویش می‌آید — وگرنه DER آن را عددِ منفی می‌خواند.
   */
  fun rawToDer(raw: ByteArray): ByteArray {
    require(raw.size % 2 == 0 && raw.isNotEmpty()) { "طولِ امضا درست نیست" }
    val half = raw.size / 2
    val r = asInteger(raw.copyOfRange(0, half))
    val s = asInteger(raw.copyOfRange(half, raw.size))

    val body = r + s
    val out = ArrayList<Byte>(body.size + 4)
    out.add(0x30)
    if (body.size >= 0x80) {
      out.add(0x81.toByte())
      out.add(body.size.toByte())
    } else {
      out.add(body.size.toByte())
    }
    out.addAll(body.toList())
    return out.toByteArray()
  }

  private fun asInteger(value: ByteArray): ByteArray {
    var start = 0
    while (start < value.size - 1 && value[start] == 0.toByte()) start++
    var trimmed = value.copyOfRange(start, value.size)
    // بیتِ اولِ روشن یعنی «منفی» در DER، پس یک صفر جلویش می‌آید
    if (trimmed.isNotEmpty() && (trimmed[0].toInt() and 0x80) != 0) {
      trimmed = byteArrayOf(0) + trimmed
    }
    return byteArrayOf(0x02, trimmed.size.toByte()) + trimmed
  }
}
